sqlmesh/core/macros.py

```diff
@@ -648,6 +649,228 @@ def eval_(evaluator: MacroEvaluator, condition: exp.Condition) -> t.Any:
+    """
+    default_null_value = exp.Literal.string("_sqlmesh_surrogate_key_null_")
+    string_fields: t.List[exp.Expression] = []
+    for i, field in enumerate(fields):
+        if i > 0:
+            string_fields.append(exp.Literal.string("|"))
+        string_fields.append(
+            exp.func(
+                "COALESCE",
+                exp.cast(field, exp.DataType.build("string")),
+                default_null_value,
+            )
+        )
+    return exp.func("MD5", exp.func("CONCAT", *string_fields))
+
+
+@macro()
+def safe_add(_: MacroEvaluator, *fields: exp.Column) -> exp.Case:
+    """Adds numbers together, substitutes nulls for 0s and only returns null if all fields are null.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @SAFE_ADD(a, b) FROM foo"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT CASE WHEN a IS NULL AND b IS NULL THEN NULL ELSE COALESCE(a, 0) + COALESCE(b, 0) END FROM foo'
+    """
+    null_cond = exp.and_(*[field.is_(exp.null()) for field in fields])
+    case = exp.Case().when(null_cond, exp.null())
+    terms: t.List[exp.Func | exp.Add] = []
+    for field in fields:
+        terms.append(exp.func("COALESCE", field, 0))
+    return case.else_(reduce(lambda a, b: a + b, terms))
+
+
+@macro()
+def safe_sub(_: MacroEvaluator, *fields: exp.Expression) -> exp.Case:
+    """Subtract numbers, substitutes nulls for 0s and only returns null if all fields are null.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @SAFE_SUB(a, b) FROM foo"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT CASE WHEN a IS NULL AND b IS NULL THEN NULL ELSE COALESCE(a, 0) - COALESCE(b, 0) END FROM foo'
+    """
+    null_cond = exp.and_(*[field.is_(exp.null()) for field in fields])
+    case = exp.Case().when(null_cond, exp.null())
+    terms: t.List[exp.Func | exp.Sub] = []
+    for field in fields:
+        terms.append(exp.func("COALESCE", field, 0))
+    return case.else_(reduce(lambda a, b: a - b, terms))
+
+
+@macro()
+def safe_div(_: MacroEvaluator, numerator: exp.Expression, denominator: exp.Expression) -> exp.Div:
+    """Divides numbers, returns null if the denominator is 0.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @SAFE_DIV(a, b) FROM foo"
+        >>> MacroEvaluator().transform(parse_one(sql)).sql()
+        'SELECT a / CASE WHEN b = 0 THEN NULL ELSE b END FROM foo'
+    """
+    return numerator / exp.Case().when(denominator.eq(0), exp.null()).else_(denominator)
+
+
+@macro()
+def union(evaluator: MacroEvaluator, *tables: exp.Table) -> exp.Union:
```

Comment:
i'm not sure about this one...

Reply:
oh, this one was actually one i was pretty happy with functionally as a net-positive value add. its pretty powerful. can you explain the issue

Reply:
i thought the code, implementation, use cases, etc on this one was probably my favorite of what i have implemented. its resilient to columns appearing in different orders, differing columns entirely, differing column types, etc. all total PITAs in sql alone. and this is well like enough in dbt. the alternative is a lot of repeated code, sources of truth in 2 places. the plan apply flow guarantees deterministic behavior. only if a column is added of the same type the the tables being unioned and becomes part of the superset would it be added to the projection (via standard plan + apply)

it doesnt stop you from just using unions regularly when you need em.

Reply:
but curious on thoughts

Reply:
Think about the scenario you have N tables which are essentially the same but come from multiple sources upstream. You want to union them all in the data warehouse to create a base table to select from. By using this macro, you gain a number of benefits. 


- If an upstream model gets a new column but the others do not, it doesn't break where a select * would
- If columns are out of order, it doesnt break where a select * would
- You don't need to repeat information already defined in your external models, aka enumerate all cols to circumvent points 1 & 2 above
- The columns are still managed explicitly via code change _somewhere_, there is nothing dynamic happening here
- When all the upstream tables get a new column of the same type, it is brought downstream


So maybe just add docs on when or why you would use it.